### DISCUSSION TITLE
Tips and Tricks to reduce the delay

### DISCUSSION THREAD

**Discussion by marcosvrs** [awawa-dev/HyperHDR#72] [General]
I couldn't find much about it, and I would love to grab the most Tips and Tricks to reduce the delay from the grabber and image processing.
For my case I have the ezcap269, running HyperHDR on my NUC7i5 and for the LED hardware I'm using the HyperSerialWLED @2000000 on a NodeMCU v3. No Smoothing, no Blackbar detection.
I see a delay of a bit less than 1s.

So, to start the list of what to do:

- Run HyperSerialEsp8266 @2000000;
- Disable Smoothing and Blackbar Detection;
- Check your hardware if you have a high usage, which can create the bottleneck;

**Comment by awawa-dev** [comment]
Something is definitely not right. Please send the full logs from HyperHDR after 2-3 minutes of working (for example put them on pastebin.com or other similar service). And in the system task manager how it looks: CPU and memory usage? There is no other heavy application running in the background?

**Reply by marcosvrs** [reply]
The latest logs: https://pastebin.com/NwAkXHCg
The current hardware usage: https://pasteboard.co/K5gEziX.png
I run it in docker, which runs 9 other containers, but my CPU and memory usages are less than 15% normally.

**Reply by awawa-dev** [reply]
I meant 2-3 minutes after the start because there some important things are happening ;) Anyway as you can see we can exclude at least video processing (2ms). BTW what are your statistics FPS or refresh rate (hz) in WLED www panel? 
I never used docker for running HyperHDR..it has a direct, exclusive connection to the serial port or it's buffered or can be shared with other containers or master system?

**Reply by marcosvrs** [reply]
Sorry, misunderstood. The log of the first minutes: https://pastebin.com/DyhyQjh6
Where can I find the FPS and HZ statistics on WLED?
The docker containers access the devices (Serial Port) directly, without buffer or any other "middleware".

**Reply by awawa-dev** [reply]
Oh, now there are too short, only half minute ;) OK, nevermind. That HyperHDR version has inferior logging that will be improved in the future release but it's seems that the smoothing is enabled?

> Where can I find the FPS and HZ statistics on WLED?

I dont remember exactly because I use HyperSerial8266. Maybe in the WLED system information or something. But it must be version that is based on WLED 0.12.
Of course take a reading with working HyperHDR.

**Reply by marcosvrs** [reply]
I recorded at 240FPS to display what I meant:
https://youtu.be/O109go6uW7o

**Reply by marcosvrs** [reply]
Regarding the Logs, it didn't change after what I sent to you... It basically duplicates the log:
`2021-06-05T23:39:33.345Z [V4L2:/DEV/VIDEO0] (INFO) Video FPS: 30.02, av. delay: 2ms, good: 1801, bad: 0 (60.03,15)`

> That HyperHDR version has inferior logging that will be improved in the future release but it's seems that the smoothing is enabled?

Smoothing is disabled (at least via UI)

I might flash the NodeMCU with the Serial version instead of WLED. Will try it later.

**Reply by marcosvrs** [reply]
WLED Info: https://pasteboard.co/K5gX53I.png

**Reply by awawa-dev** [reply]
> https://youtu.be/O109go6uW7o
But 1sec in the movie it's not real 1sec? Or it is? I dont know to assess that..maybe the LED refresh rate is limited. And what are your real expectation? Ezcap 269 has a typical delay around 300ms, I dont how how it's connected because sometimes some other devices on the path can introduce additional lag. For comparison Ezcap 320 has 50ms delay. I dont know if it's related...For comparision from the blog https://streamable.com/zutven (Ezcap 320)


First please take a reading from WLED  (it was introduced in WLED 0.12 so be sure that you are using the latest version HyperserialWLED). In HyperSerial8266 it's a bit more difficult to read the FPS and they are not live statistics. 

Also I would test a HyperHDR running on the main system to exclude some problems related to the docker because that setup is not officially supported (don't have it in the test paths).

**Reply by awawa-dev** [reply]
> WLED Info: https://pasteboard.co/K5gX53I.png

That's bad. 

One more test:
Enable Smoothing with refresh rate for example 80hz.
Enable continues output in the the smoothing.
Than again take a reading from the WLED with HyperHDR running.

**Reply by marcosvrs** [reply]
> Ezcap 269 has a typical delay around 300ms

That explains a lot. The delay in total is around 400~800ms.

The video is at 30FPS, but the recording was 240FPS. Which means that 1s ~ 300ms.

**Reply by marcosvrs** [reply]
With Smoothing at 80Hz and Continues Output: WLED report 21FPS
https://pasteboard.co/K5h9UuX.png

**Reply by awawa-dev** [reply]
> With Smoothing at 80Hz and Continues Output: WLED report 21FPS

That's the problem: you have got a bottleneck probably at the communication with the LED strip. With a real 21FPS led output and 30FPS from the grabber at some stage it will overflow or the data will be buffered and render with a delay. What's the LED strip type by the way? I doubt it can be a problem because even sk6812 rgbw should be able to render around 70FPS on the hardware level.

Another test and possible workaround:
- disable again smoothing
- in the video grabber configuration set frame rate decimation to 2 
- take a reading from the WLED: should be around 15FPS (30FPS from the grabber divided by framedecimtion)
- assess if the lag is reduced using again some test-video

**Reply by marcosvrs** [reply]
To compare, I recorded all three suggestions that you've made:
No Smoothing, no Blackbar, no frame skipping: https://streamable.com/y8gm8d
Only Frame Skipping: https://streamable.com/2mq7h2
Only smoothing at 80Hz and continues output: https://streamable.com/9fv6wn

With the frame skipping, I can see 15FPS on WLED: https://pasteboard.co/K5lLXGj.png

**Reply by marcosvrs** [reply]
About my hardware:
- SK9822 60LED/m with 263 LEDs
- NodeMCU v3 running HyperSerialWLED
- Power Supply 5V 20A
- Ezcap269
- NUC7i5 connected with Ezcap269 via USB 3.0 and NodeMCU v3 via USB 2.0

HDMI connection flow:
- Apple TV
- HTR-4071 - Yamaha AV Receiver
- Ezcap269
- TCL-65DP676

**Reply by awawa-dev** [reply]
And how its connected: from which device to the other?

From some reason you are capped to around 20-23FPS on the ESP device despite 2Mb serial port speed.
> With Smoothing at 80Hz and Continues Output: WLED report 21FPS

'Continues output' is to some extended independent from the rate from the grabber and takes very little resource so they are not a problem at all. It forces constant frame-rate so with this setting you should have around 70-80FPS in the WLED no matter what happens. At least when there is no other bottleneck outside the system for example the led driver is not capable of doing that from some reason...

There are two options that come to my head:
- docker is limiting the transfer or there is some conflict of shared resources between some other container or system. The transfer is buffered and it's not working well resulting in ~20FPS.
- there is some limitation from the WLED, some of its settings that caps the refresh rate. At this point installing HyperSerialESP8266 could provide more information because its statistics are independent from the LED driver and are related to serial port communication only.

**Reply by marcosvrs** [reply]
TBH, I'm reluctant about installing it directly on my host system, but I'll do that if the HyperSerialESP8266 doesn't help.
I'm currently flashing my NodeMCU with HyperSerialESP8266.

**Reply by awawa-dev** [reply]
If you install precompiled version of HyperSerialESP8266 the led strip won't work because they are meant for ws8212/13/sk6812 and similar one communication line led strip. But what is important that it will show the statistics from the serial port.

**Comment by marcosvrs** [comment]
Is it normal that the first LED of the strip is red when no color is displayed?

**Reply by awawa-dev** [reply]
Answered in the previous topic also so check it please.

Did you play with WLED settings? Maybe skip first led option is enabled or your led strip definition in HyperHDR is invalid (but really don't know how to achieve that effect intentionally).
